fix(compiler): ensure that partially compiled queries can handle forward references

petebacondarwin opened the pull request:
When a partially compiled component or directive is "linked" in JIT mode, the body
of its declaration is evaluated by the JavaScript runtime. If a class is referenced
in a query (e.g. `ViewQuery` or `ContentQuery`) but its definition is later in the
file, then the reference must be wrapped in a `forwardRef()` call.

Previously, query predicates were not wrapped correctly in partial declarations
causing the code to crash at runtime. In AOT mode, this code is never evaluated
but instead transformed as part of the build, so this bug did not become apparent
until Angular Material started running JIT mode tests on its distributable output.

This change fixes this problem by noting when queries are wrapped in `forwardRef()`
calls and ensuring that this gets passed through to partial compilation declarations
and then suitably stripped during linking.

See https://github.com/angular/components/pull/23882 and https://github.com/angular/components/issues/23907

**Unfortunately, any libraries that were published before this gets merged will potentially have the problem going forward, if they have forwardRefs in their query definitions. The resolution it to simply build and release a new version. With that in mind, it might be worth landing this in 12.2.x LTS as well.**

petebacondarwin committed:
fix(compiler): ensure that partially compiled queries can handle forward references

When a partially compiled component or directive is "linked" in JIT mode, the body
of its declaration is evaluated by the JavaScript runtime. If a class is referenced
in a query (e.g. `ViewQuery` or `ContentQuery`) but its definition is later in the
file, then the reference must be wrapped in a `forwardRef()` call.

Previously, query predicates were not wrapped correctly in partial declarations
causing the code to crash at runtime. In AOT mode, this code is never evaluated
but instead transformed as part of the build, so this bug did not become apparent
until Angular Material started running JIT mode tests on its distributable output.

This change fixes this problem by noting when queries are wrapped in `forwardRef()`
calls and ensuring that this gets passed through to partial compilation declarations
and then suitably stripped during linking.

See https://github.com/angular/components/pull/23882 and https://github.com/angular/components/issues/23907

alxhub commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -233,11 +233,17 @@ export interface R3DeclareQueryMetadata {
+  /**
+   * True if the `predicate` is wrapped in a `forwardRef()` invocation.
+   */
+  isForwardRef: boolean;
```
👍 that seems okay for now, but we should definitely rename `R3ProviderExpression` to something more appropriate.

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/queries/view_query_forward_ref.js:
```diff
@@ -0,0 +1,16 @@
+  selectors: [["view-query-component"]],
+  viewQuery: function ViewQueryComponent_Query(rf, ctx) {
+    if (rf & 1) {
+      $r3$.ɵɵviewQuery(SomeDirective, 5);
```
This could use `__QueryFlags.descendants__|__QueryFlags.emitDistinctChangesOnly__` as the flags are supported as verification macros.

petebacondarwin replied:
I totally forgot about these. Thank goodness for you @JoostK !

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/queries/content_query_forward_ref.js:
```diff
@@ -0,0 +1,27 @@
+  selectors: [["content-query-component"]],
+  contentQueries: function ContentQueryComponent_ContentQueries(rf, ctx, dirIndex) {
+    if (rf & 1) {
+    $r3$.ɵɵcontentQuery(dirIndex, SomeDirective, 5);
```
Here as well for the flags. Also, it appears that the indentation is off here (as is the `let $tmp$;` statement below)

petebacondarwin committed:
refactor(compiler): rename R3ProviderExpression and associated helpers

This interface will be used in other situations so this change renames it to be more general as `MaybeForwardRefExpression`.

michaelfaith commented:
I wonder if the issue this is addressing would cause the below errors.  I built this library with the 13.0.0 compiler, and similar to Material, these errors are popping up in unit tests that involve components from the library, but it runs fine with `ng serve` (aot).  Different errors than the Material ones, but similarly related to the shape of the compiled library.   So I thought it might be related to this.  If it isn't, then I'll have another issue to track down. 

```
Details:

    C:\Users\<user>\src\fabric-starter\node_modules\@fabric\components\fesm2015\fabric-components-breadcrumbs.mjs:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){import * as i4 from '@angular/common';
                                                                                      ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      10 | import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
      11 | import { RouterTestingModule } from '@angular/router/testing';
    > 12 | import { FabBreadcrumbsModule } from '@fabric/components/breadcrumbs';
         | ^
      13 | import { FabHeaderModule } from '@fabric/components/header';

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1728:14)
      at Object.<anonymous> (src/app/app.component.spec.ts:12:1)

```

and 

```
 Details:

    C:\Users\<user>\src\fabric-starter\node_modules\@fabric\components\fesm2015\fabric-components-utilities.mjs:196
    export { compareById, convertDateTimeToTimestamp, doesContainOnlyStrings, findObjWithKey, getCoreRoute, isNotNull, isNotNullOrUndefined, isNull, isNullOrUndefined, isObject, isUndefined, removeIntersection, removeItemFromArray,
removeLeadingSlash, removeQueryParams, replaceAll };
    ^^^^^^

    SyntaxError: Unexpected token 'export'

       7 | import { Injectable } from '@angular/core';
    >  8 | import { isNotNull } from '@fabric/components/utilities';
         | ^
        9 | import { AuthenticationService } from '../authentication/authentication.service';
      10 |
      11 | /**

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1728:14)
      at Object.<anonymous> (src/app/core/user/user.service.ts:9:1)

```

JoostK commented:
@michaelfaith that does not look related to what is being fixed here, it seems. I believe that Jest creates a CommonJS wrapper for any module it loads, in order to sandbox things like `require` and `module` and the like. That does require however that CommonJS code is used, whereas it appears that in your case the ESM ends up within the CommonJS wrapper. That is invalid, as ESM must be top-level.

JoostK commented:
You can actually spot the CJS wrapper in the error message:

```ts
function(module,exports,require,__dirname,__filename,jest){
```

michaelfaith commented:
@JoostK appreciate the info.  That helps point me in the right direction.

JoostK commented on packages/core/test/render3/jit/declare_directive_spec.ts:
```diff
@@ -118,6 +118,27 @@ describe('directive declaration jit compilation', () => {
   });
 
+  it('should compile content queries with forwardRefs', () => {
+    class Child {}
```
Please move `Child` below `def`, to make it a true forward ref.

petebacondarwin replied:
Doh! I did have it below but then was testing it with and without the forwardRef so ended up moving it above. Good spot. Thanks

JoostK commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -306,15 +306,21 @@ function convertQueryDeclarationToMetadata(declaration: R3DeclareQueryMetadataFa
+    string[] {
+  return Array.isArray(predicate) ?
+      predicate :
+      createMayBeForwardRefExpression(new WrappedNodeExpr(resolveForwardRef(predicate)), false);
```
Can we be certain that the forward ref is available by the time we unwrap it here? I'm afraid we can't, as this is called immediately in `ɵɵngDeclareDirective` _before_ having reached the declaration point of what's inside the forward ref.

petebacondarwin replied:
And indeed fixing the JIT test causes it to fail and indicate that you are correct here.

JoostK commented on packages/compiler/src/injectable_compiler_2.ts:
```diff
@@ -118,8 +118,9 @@ export function compileInjectable(
         'providedIn',
-        meta.providedIn.isForwardRef ? generateForwardRef(meta.providedIn.expression) :
-                                       meta.providedIn.expression);
+        meta.providedIn.forwardRef === ForwardRefHandling.Unwrapped ?
```
This looks like it can use `convertFromMaybeForwardRefExpression`.

JoostK reviewed (approved):
In general LGTM, but I'm deferring for a full-pass review to @alxhub.

atscott commented:
[presubmit](http://tap/408898616)

atscott commented:
This PR was merged into the repository by commit 393efa54e669c4ae1e479fd59fc4f9ea69ab77e9.
